File: LeetCode/Medium/2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        vector<TreeNode*> nodes(100001);
        unordered_set<int> child;
        int ret = 0;

        for (const auto& v : descriptions) {
            int p = v[0];
            int c = v[1];

            if (!nodes[p]) nodes[p] = new TreeNode(p);
            if (!nodes[c]) nodes[c] = new TreeNode(c);

            child.insert(c);

            if (v[2] == 1) nodes[p]->left = nodes[c];
            else nodes[p]->right = nodes[c];
        }

        for (const auto& v : descriptions) {
            if (child.find(v[0]) == child.end()) {
                ret = v[0];
                break;
            }
        }

        return nodes[ret];
    }
};

File: LeetCode/Medium/2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        nodes = {}
        child = set()
        ret = None

        for v0, v1, v2 in descriptions:
            if v0 not in nodes:
                nodes[v0] = TreeNode(v0)
            if v1 not in nodes:
                nodes[v1] = TreeNode(v1)
            
            child.add(v1)

            if v2:
                nodes[v0].left = nodes[v1]
            else:
                nodes[v0].right = nodes[v1]
        
        for k in nodes.keys():
            if k not in child:
                ret = k
                break

        return nodes[ret]